### ISSUE TITLE
Investigate further chunk compression.

### ISSUE THREAD

**Issue by krasi-georgiev** [prometheus/tsdb#598]
https://groups.google.com/forum/#!msg/prometheus-users/I7uneXEisV0/_BIMTqpWAAAJ

The thread suggests that even a single chunk can be compressed by more than 50% so worth investigating.

**Comment by beorn7** [comment]
That's interesting indeed.

Some fuzzy data points from the depths of my memories:

In Prometheus v1 (using the chunk encoding v2, AKA Gorilla-style double delta), I remember having played with compressing the series files. And indeed, they compressed very little, if at all, telling me that we were already quite close to the max-entropy optimum (or whatever that's called in information theory). However, I should note that no real research went into the details of the chunk encoding. I did it all _ad hoc_, following my gut feeling what might work well. And apparently, things worked really well.

Rather than refining or researching the chunk encoding, the whole storage got a rewrite. At its core, samples in the chunks are still using Gorilla-style double delta encoding, albeit in a “purer” form, without some of the addition I had added to the old storage following my gut feeling. I don't know how much research went into that. I haven't seen any documentation about it and don't remember any personal communication. Since in general, the observed compression with Prometheus v2 looked even better than in Prometheus v1, I assumed my added encoding ideas in v1 weren't very relevant.

However, when finally converting long-term Prometheus servers from v1 to v2 at SoundCloud, I saw some special cases where v2 had significant worse compression behavior than v1, making me think again about the differences. Semi-educated guesses:
- v1 will have worse compression under series churn because it has a fixed chunk size, i.e. series churn will create a lot of incompletely filled chunks, wasting space.
- v1 uses LevelDB for indices, while v2 has custom and integrated indexing. It's fair to assume that the v1 indexing is way less efficient in terms of space, and this will again be more relevant under series churn.
- v2 in turn repeats the index for every block, thus losing efficiency if the index is essentially the same for each block (if there is no series churn), which gets more relevant the better the sample data alone compresses (as the index is relatively larger then).

From the above, v1 would be most efficient in cases with very few but very long lasting series. And that's interestingly exactly the case where I saw v2 taking more disk space.

Again, this is all fuzzy memory and no real research, which would be required to confirm the above. For the record, here are the “add-ons” I had implemented for v1 that exist neither in vanilla-Gorilla (rhymes!) nor in Prometheus v2. They are all based on guesses of the nature of the samples.

- If the samples look like random numbers, no double delta encoding is used at all. The samples are saved as is as every attempt of compression will just be overhead in terms of both computation and space.
- If the samples look like integers rather than floats, they are treated similar to the time stamps, i.e. encode them as integers with straight-forward double-deltas rather than the more complex operations otherwise done on the bit-patterns of a float.
- If the samples are mostly constant (like the ubiquitous `up==1` or an info metric), what's encoded is essentially “repeat the value `1` 98 times”.

Perhaps tricks like this could benefit v2. In v1, an encoding was picked speculatively, and if new samples wouldn't fulfill the requirement (suddenly not constant anymore or not integer), the whole chunk got transcoded. This approach is not possible with v2 anymore by design. When I asked Fabian why he kicked out that kind of flexibility (and the other encodings in general), he said their effectiveness could be investigated later. If needed, the change of encoding could be included at the block compaction step, i.e. always start with vanilla-Gorilla as we have it now and then analyze the data during compaction, picking other encodings for series that are only integers or even constant. That might work just great.

**Comment by krasi-georgiev** [comment]
Thanks for the info , that is quite useful. 
Especially the part for analyzing data and applying some additional encoding/compression during compaction sound like a very good idea.

**Comment by brian-brazil** [comment]
We're not doing any compression of the postings tables, which is probably where there's most to gain in the indexes. A simple delta compression would probably buy a good bit - but would make Seeking slower.

**Comment by bboreham** [comment]
Anecdote from our Cortex installation: changing from v1 `varbit` to v2 `xor` the data got something like 20% bigger.  This was balanced by a simultaneous change which means we have 30% fewer chunks, but I'd certainly be in favour of work to regain the previous level of compression.

I've also wondered if [this constant](https://github.com/prometheus/tsdb/blob/6ac81cc7a9c5b1684d0cff71a57dff1beea9f898/head.go#L1637) is optimal for Prometheus data.

**Comment by dipack95** [comment]
I've been looking at possibly compressing the time series chunks for prometheus, and the WAL compression PR #609 got me thinking that a snappy buffer for the chunks data would work for my use case. I've got a quick and dirty version that writes and reads as expected, except for compaction.

```
272K    ./uncompressed_chunk/chunks
272K    total

152K    ./snappy_compressed_chunk/chunks
152K    total
```

After reading this issue, I've come to understand that the project is looking to do something similar, except the compression algorithm chosen would depend on the kind of time series chunk data that it would be compressing, instead of a blanket compression approach like mine. Would it make sense for me to continue down this path?

**Comment by krasi-georgiev** [comment]
I guess some research and testing would be required to decide between both approaches. Knowing the benefits and trade-offs  can help take the decision about the design.

We can run  benchmark test using prombench and see the results for each approach.

**Comment by krasi-georgiev** [comment]
maybe starting with a proposal doc would be a good start so that more people can share ideas and experience.

**Comment by dipack95** [comment]
@krasi-georgiev I'm new to this process, so by proposal doc, do you mean a new GitHub issue or PR?

**Comment by brian-brazil** [comment]
A Google doc shared with -developers would be the usual approach for something as undefined as this.

**Comment by krasi-georgiev** [comment]
@dipack95 also be prepared with some patient as this is not going to be an easy task.
I imagine that just the proposal doc and research would probably take a bit of time.

For someone to take up in this task would require quite a bit of dedication and patience.

**Comment by dipack95** [comment]
@krasi-georgiev Yes, I understand that such a proposal, with such wide impact will take some time to flesh out and fully implement. This feature, if implemented, could make a big difference in our prometheus deployments, and thus I thought it wise to contact the team to see if I could help.

I've create a google doc here: https://docs.google.com/document/d/1kujMhPwWdW5BpdrJpiq-CEcD3M58yk755LkqSLH3dL0/edit?usp=sharing
Please tell me if you can access it!

**Comment by krasi-georgiev** [comment]
thanks!

nope I can't access. Best to make it open for viewing and commenting.

**Comment by dipack95** [comment]
Here's a new link, which should be more permissive https://docs.google.com/document/d/1Je2DmMdvqQlWuXghGxkbgRdnNzoKYSeIkUQDWZ-GMcs/edit?usp=sharing

**Comment by krasi-georgiev** [comment]
this one is open for public writing.
I would say to set - can comment, can view .
